### ISSUE TITLE
Glob Pattern Entities Not Resolved

### ISSUE THREAD

**Issue by Adrastopoulos** [typeorm/typeorm#9004]
## Issue Description

The Directory Exported Classes Loader  (`DirectoryExportedClassesLoader.js`) normalizes the path provided to the `entities` property on a `DataSourceOptions` object. This normalized path changes all **forward** slashes to **back** slashes - however, glob is only compatible with **forward** slashes. Removing the `pathNormalize` function fixed my issue.

### Expected Behavior

```
All classes found using provided glob pattern...
```

### Actual Behavior

```
 No classes were found using the provided glob pattern...
```

### Steps to Reproduce

1. Replicate Code Below

```typescript
const databaseOptions: DataSourceOptions = {
	type: 'postgres',
	host: 'localhost',
	port: 5432,
	username: 'postgres',
	password: DB_PASSWORD,
	entities: [path.join(__dirname, '/entities/*.js').replace(/\\/g, '/')],
	database: 'postgres',
	logging: ['info'],
};

const dataSource = await new DataSource(databaseOptions).initialize();

console.log(dataSource.entityMetadatas); // []
```

### My Environment

| Dependency          | Version  |
| ---                 | ---      |
| Operating System    |     Windows     |
| Node.js version     | 8.3.0  | 
| Typescript version  | 4.6.4  | 
| TypeORM version     | 0.3.6  | 

### Relevant Database Driver(s)

| DB Type           | Reproducible  |
|-------------------| ---           |
| `aurora-mysql`    | no            |
| `aurora-postgres` | no            |
| `better-sqlite3`  | no            |
| `cockroachdb`     | no            |
| `cordova`         | no            |
| `expo`            | no            |
| `mongodb`         | no            |
| `mysql`           | no            |
| `nativescript`    | no            |
| `oracle`          | no            |
| `postgres`        | yes            |
| `react-native`    | no            |
| `sap`             | no            |
| `spanner`         | no            |
| `sqlite`          | no            |
| `sqlite-abstract` | no            |
| `sqljs`           | no            |
| `sqlserver`       | no            |

### Are you willing to resolve this issue by submitting a Pull Request?

 - ✅ Yes, I have the time, and I know how to start.
 - ✖️ Yes, I have the time, but I don't know how to start. I would need guidance.
 - ✖️ No, I don’t have the time, but I can support (using donations) development.
 - ✖️ No, I don’t have the time and I’m okay to wait for the community / maintainers to resolve this issue.

**Comment by avin** [comment]
I have same problem on windows and there are no troubles in unix environment with same configuration. 
Eventually had to move to WSL. :(

**Comment by avin** [comment]
Found the soultion. Bug is in "glob" package. You have to downgrade "glob" dependency to 7.2.0 for typeorm and it starts to work. [(how to change it)](https://stackoverflow.com/questions/15806152/how-do-i-override-nested-npm-dependency-versions)

**Comment by Adrastopoulos** [comment]
Alternatively, upgrading the glob dep to `7.2.3` fixed the issue as well. Thanks for the PR.

**Comment by szul** [comment]
I spent all weekend trying to figure out why an entire application environment stopped working. Today I was debugging TypeORM all the way down to `DirectoryExportedClassesLoader.js` and hit the same issue. After trying various glob patterns, I stuck in a full string and used the debug window in VSCode directly on the server. It worked before `pathNormalize()`, but not after. We are not yet on NPM 8, so I'm going to attempt a different path to downgrade.

**Comment by lidroider** [comment]
I have faced same problem on typeorm 0.3.12 and windows 10. But on linux or mac it's working normally

I fixed by overriding `glob` -> 7.2.3 in `package.json` like 
```
"overrides": {
  "glob": "7.2.3"
}
```

**Comment by ownik** [comment]
Yep. I also have the same problem on typeorm 0.3.12 and windows 11. Overrides helps me too

**Comment by Vilx2** [comment]
+1 to this. Just spent half a day tearing my hairs out about this. Thanks to @lidroider for the "overrides" trick, I didn't know that.

That said, I think the bug is on TypeORM side of things. Glob/minimatch explicitly state in the READMEs:

> **Please only use forward-slashes in glob expressions.**
>
> Though windows uses either / or \ as its path separator, only / characters are used by this glob implementation. You must use forward-slashes **only** in glob expressions. Back-slashes in patterns will always be interpreted as escape characters, not path separators.

However TypeORM explicitly calls `PathNormalize()` which converts all `/` to `­\`. This is wrong.

**Comment by ownik** [comment]
I guess it's already resolved in PR #9768 3 weeks ago. But unfortunately, it's not included in 0.3.12

**Comment by ownik** [comment]
I checked `0.3.13-dev.58fc088` and it includes fix and works fine.

**Comment by akmaldira** [comment]
> 

Thanks you, i stess about this issue in 3 days

**Comment by cooljeffro** [comment]
> I have faced same problem on typeorm 0.3.12 and windows 10. But on linux or mac it's working normally
> 
> I fixed by overriding `glob` -> 7.2.3 in `package.json` like
> 
> ```
> "overrides": {
>   "glob": "7.2.3"
> }
> ```

This also worked for me, but I only overrode typeorm's glob version.  I also had to uninstall and reinstall typeorm for the overriden glob version to get installed.
```
"overrides": {
  "typeorm": {
    "glob": "7.2.3"
  }
}
```

**Comment by mfuatnuroglu** [comment]
Can't believe that I spent hours on this problem.
I replaced the version of the `typeorm` in the `package.json` file with `"typeorm": "0.3.13-dev.58fc088"`

**Comment by foaudkajj** [comment]
Can't believe that I spent hours on this problem.
I replaced the version of the `typeorm` in the `package.json` file with `"typeorm": "0.3.13-dev.58fc088"`

see https://github.com/typeorm/typeorm/pull/9768

**Comment by KevinUre** [comment]
i put `"overrides": {  "glob": "7.2.3" },` in my package.json, uninstalled typeorm, deleted my package-lock.json. deleted node_modules, reinstalled packages, then re-added typeorm and it didn't use the specified version
```
> npm list glob
project@0.0.1
├─┬ @nestjs/cli@9.2.0
│ └─┬ shelljs@0.8.5
│   └── glob@7.2.3
├─┬ eslint@8.36.0
│ └─┬ file-entry-cache@6.0.1
│   └─┬ flat-cache@3.0.4
│     └─┬ rimraf@3.0.2
│       └── glob@7.2.3 deduped
├─┬ jest@29.3.1
│ └─┬ @jest/core@29.5.0
│   ├─┬ @jest/reporters@29.5.0
│   │ └── glob@7.2.3 deduped
│   ├─┬ jest-config@29.5.0
│   │ └── glob@7.2.3 deduped
│   └─┬ jest-runtime@29.5.0
│     └── glob@7.2.3 deduped
├─┬ ts-jest@29.0.3
│ └─┬ babel-jest@29.5.0
│   └─┬ babel-plugin-istanbul@6.1.1
│     └─┬ test-exclude@6.0.0
│       └── glob@7.2.3 deduped
└─┬ typeorm@0.3.12
  └── glob@8.1.0
```

i tried this version too:
```
"overrides": {
  "typeorm": {
    "glob": "7.2.3"
  }
}
```
what am i missing here?

**Comment by Samandar02** [comment]
```datasource.ts

import { join } from "node:path";
import { DataSource } from "typeorm";

export let dataSource = new DataSource({
    type: 'postgres',
    host: 'localhost',
    port: 5432,
    database: 'localdb',
    username: 'admin',
    password: 'admin',
    schema: 'public',
    entities: [join(__dirname, '/entities/*.ts').replace(/\\/g, '/')],
    migrations: [join(__dirname, '/migrations/*.ts').replace(/\\/g, '/')],
})
```
this snipped code works for me width Windows,
> note: `datasource.ts` file, `migrations` and `entities` folder inside of same floder `db`

**Comment by fortunaphantom** [comment]
I was facing `TypeORMError: Entity metadata for xxx#xx was not found. Check if you specified a correct entity object and if it's connected in the connection options` in typeorm@0.3.10 and nestjs. It worked well on ubuntu but not worked on windows.

Updating package.json resolved this issue.
"overrides": {
    "glob": "7.2.3"
  }

Thank you to @lidroider

**Comment by Vilx2** [comment]
Double checked this. Seems like issue #9766 was a duplicate of this bug and got fixed over 2 years ago. So all is good. :)
